### PR TITLE
[论文复现赛] FCCDN

### PR DESCRIPTION
**Pull request by liuxtakeoff**
# 运行结果总览：

# tipc全过程：

### CODE REVIEW

File: tutorials/train/change_detection/fccdn.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    file_list=EVAL_FILE_LIST_PATH,
+    label_list=None,
+    transforms=eval_transforms,
+    num_workers=0,
+    shuffle=False,
+    with_seg_labels=False,
+    binarize_labels=True)
+
+# 使用默认参数构建FCCDN模型
+# 目前已支持的模型请参考：https://github.com/PaddlePaddle/PaddleRS/blob/develop/docs/apis/model_zoo.md
```
Comment (Bobholamovic) [inline tutorials/train/change_detection/fccdn.py]: 无效的URL

**Review by Bobholamovic — CHANGES_REQUESTED** [review]
我认为bcd loss和普通的交叉熵应该解耦，也即模型在training模式下应该有两个输出：第一个输出是最终的logits，用来计算交叉熵；第二个是用来计算自监督损失所需要的输入。这样设计之后会方便用户进行loss组合。

**Comment by liuxtakeoff** [comment]
已按照要求对损失函数进行了解耦，并修改了链接及描述。

File: paddlers/rs_models/cd/fccdn.py [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+class FCCDN(nn.Layer):
+    """
+    The FCCDN implementation based on PaddlePaddle.
+
+    The original article refers to
+        Pan Chen, et al., "FCCDN: Feature Constraint Network for VHR Image Change Detection"
+        (https://arxiv.org/pdf/2105.10860.pdf).
+
+    Args:
+        in_channels(int): Number of input channels(default: 3).
```
Comment (Bobholamovic) [inline paddlers/rs_models/cd/fccdn.py]: 括号和单词之间需要1个空格。

File: paddlers/rs_models/cd/fccdn.py [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+
+        y1 = self.conv1(y)
+        y2 = self.conv2(y1)
+        y3 = self.conv3(y2 + y1)
+        out = self.conv_out(paddle.abs(x1 + x2 + x3 - y1 - y2 - y3))
+        return out
+
+
+class DFModule(nn.Layer):
+    """dense connection-based feature fusion module"""
```
Comment (Bobholamovic) [inline paddlers/rs_models/cd/fccdn.py]: 首字母大写。

File: paddlers/rs_models/cd/fccdn.py [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+        x3 = self.conv3(x2 + x1)
+
+        y1 = self.conv1(y)
+        y2 = self.conv2(y1)
+        y3 = self.conv3(y2 + y1)
+
+        return self.conv_out(x1 + x2 + x3 + y1 + y2 + y3)
+
+
+class DensecatCatDiff(nn.Layer):
```
Comment (Bobholamovic) [inline paddlers/rs_models/cd/fccdn.py]: DensecatCatDiff还是DenseCatDiff？为什么有两个cat

File: paddlers/rs_models/cd/fccdn.py [hunk 5]
```diff
@@ -0,0 +1,478 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import paddle
+import paddle.nn as nn
+import paddle.nn.functional as F
+
+from paddlers.rs_models.cd.layers import BasicConv, MaxPool2x2, Conv1x1, Conv3x3
```
Comment (Bobholamovic) [inline paddlers/rs_models/cd/fccdn.py]: 使用相对引用，直接`from .layers import ...`即可。

File: paddlers/rs_models/cd/__init__.py [hunk 6]
```diff
@@ -23,3 +23,5 @@
 from .fc_siam_conc import FCSiamConc
 from .fc_siam_diff import FCSiamDiff
 from .changeformer import ChangeFormer
+from .fccdn import FCCDN
+from .losses import fccdn_loss_ssl
```
Comment (Bobholamovic) [inline paddlers/rs_models/cd/__init__.py]: 建议名字直接叫`ssl_loss`。

File: paddlers/rs_models/cd/losses/fccdn_loss.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+        return loss
+
+    def __call__(self, y_true, y_pred):
+        return self.soft_dice_loss(y_true, y_pred.astype(paddle.float32))
+
+
+class MultiClassDiceLoss(nn.Layer):
+    def __init__(
+            self,
+            weight: paddle.Tensor,
```
Comment (Bobholamovic) [inline paddlers/rs_models/cd/losses/fccdn_loss.py]: 如果其它的函数都没有用type hint，那我建议保持风格统一这里也不需要用。

File: paddlers/rs_models/cd/losses/fccdn_loss.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+            do_sigmoid: Optional[bool]=False,
+            **kwargs, ) -> paddle.Tensor:
+        super(MultiClassDiceLoss, self).__init__()
+        self.ignore_index = ignore_index
+        self.weight = weight
+        self.do_sigmoid = do_sigmoid
+        self.binary_diceloss = DiceLoss(batch)
+
+    def __call__(self, y_pred, y_true):
+        if self.do_sigmoid:
```
Comment (Bobholamovic) [inline paddlers/rs_models/cd/losses/fccdn_loss.py]: 为什么写的是`do_sigmoid`，实际功能却是do softmax？

File: paddlers/rs_models/cd/losses/fccdn_loss.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+            j = y_pred.sum(1).sum(1).sum(1)
+            intersection = (y_true * y_pred).sum(1).sum(1).sum(1)
+        score = (2. * intersection + smooth) / (i + j + smooth)
+        return score.mean()
+
+    def soft_dice_loss(self, y_true, y_pred):
+        loss = 1 - self.soft_dice_coeff(y_true, y_pred)
+        return loss
+
+    def __call__(self, y_true, y_pred):
```
Comment (Bobholamovic) [inline paddlers/rs_models/cd/losses/fccdn_loss.py]: 不建议直接重写`nn.Layer`对象的`__call__`方法。

File: paddlers/rs_models/cd/losses/fccdn_loss.py [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+            ignore_index: Optional[int]=-1,
+            do_sigmoid: Optional[bool]=False,
+            **kwargs, ) -> paddle.Tensor:
+        super(MultiClassDiceLoss, self).__init__()
+        self.ignore_index = ignore_index
+        self.weight = weight
+        self.do_sigmoid = do_sigmoid
+        self.binary_diceloss = DiceLoss(batch)
+
+    def __call__(self, y_pred, y_true):
```
Comment (Bobholamovic) [inline paddlers/rs_models/cd/losses/fccdn_loss.py]: 有的loss把`y_pred`放在前面、`y_true`放在后面，有的则不是。建议统一把pred放前面gt放后面。

File: paddlers/rs_models/cd/losses/fccdn_loss.py [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+        for i in range(y_pred.shape[1]):
+            if i != self.ignore_index:
+                diceloss = self.binary_diceloss(y_pred[:, i, :, :],
+                                                y_true[:, i, :, :])
+                total_loss += paddle.multiply(diceloss, self.weight[i])
+                tmp_i += 1.0
+        return total_loss / tmp_i
+
+
+class DiceBceLoss(nn.Layer):
```
Comment (Bobholamovic) [inline paddlers/rs_models/cd/losses/fccdn_loss.py]: 名字应该叫`DiceBCELoss`

File: paddlers/rs_models/cd/losses/fccdn_loss.py [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+    """Binary loss"""
```
Comment (Bobholamovic) [inline paddlers/rs_models/cd/losses/fccdn_loss.py]: 这个注释没有描述清楚类的功能，只从`binary`这个词并得不到太多信息。

File: paddlers/rs_models/cd/losses/fccdn_loss.py [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+
+    def __init__(self):
+        super(DiceBceLoss, self).__init__()
+        self.bce_loss = nn.BCELoss()
+        self.binnary_dice = DiceLoss()
+
+    def __call__(self, scores, labels, do_sigmoid=True):
+
```
Comment (Bobholamovic) [inline paddlers/rs_models/cd/losses/fccdn_loss.py]: 此处无需空行。

File: paddlers/rs_models/cd/losses/fccdn_loss.py [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+        if len(labels.shape) > 3:
+            labels = labels.squeeze(1)
+        if do_sigmoid:
+            scores = paddle.nn.functional.sigmoid(scores.clone())
+        diceloss = self.binnary_dice(scores, labels)
+        bceloss = self.bce_loss(scores, labels)
+        return diceloss + bceloss
+
+
+class McDiceBceLoss(nn.Layer):
```
Comment (Bobholamovic) [inline paddlers/rs_models/cd/losses/fccdn_loss.py]: `Bce`->`BCE`

File: paddlers/rs_models/cd/losses/fccdn_loss.py [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+    pool = paddle.nn.MaxPool2D(kernel_size=2, stride=(2, 2), padding=(0, 0))
+    labels_downsample = pool(labels_downsample)
+
+    # Seg map
+    out1 = paddle.nn.functional.sigmoid(scores[0]).clone()
+    out2 = paddle.nn.functional.sigmoid(scores[1]).clone()
+    out3 = out1.clone()
+    out4 = out2.clone()
+
+    out1[labels_downsample == 1] = 0
```
Comment (Bobholamovic) [inline paddlers/rs_models/cd/losses/fccdn_loss.py]: 注意数值类型。此时`labels_downsample`的dtype为`paddle.float32`，由于[浮点误差](https://www.zhihu.com/question/380574329)的存在，不建议直接对浮点型使用`==`、`!=`、`<=`等运算符。

File: paddlers/rs_models/cd/losses/fccdn_loss.py [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+    FCCDN change detection self-supervised learning loss
+    Args:
+        scores(list) = model(input) = [y1, y2]
+        labels(tensor-int64) = binary_cd_labels
+    """
+
+    # Create loss
+    criterion_ssl = DiceBceLoss()
+
+    # Get downsampled change map
```
Comment (Bobholamovic) [inline paddlers/rs_models/cd/losses/fccdn_loss.py:129]: 为了增强通用性，建议可以修改为将`labels` resize到`scores[0]`大小。

File: paddlers/rs_models/cd/losses/fccdn_loss.py [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+    out2 = paddle.nn.functional.sigmoid(scores[1]).clone()
+    out3 = out1.clone()
+    out4 = out2.clone()
+
+    out1[labels_downsample == 1] = 0
+    out2[labels_downsample == 1] = 0
+    out3[labels_downsample != 1] = 0
+    out4[labels_downsample != 1] = 0
+
+    pred_seg_pre_tmp1 = paddle.ones(out1.shape)
```
Comment (Bobholamovic) [inline paddlers/rs_models/cd/losses/fccdn_loss.py]: 类似的写法使用`paddle.where`实现是否更优雅一些？另外可以通过记录mask来防止每次都要重新计算二值索引用的mask。

**Review by Bobholamovic — CHANGES_REQUESTED** [review]
TIPC耗时太长，lite_train_lite_infer训练过程应该保持在10分钟以内，建议调参并减小epoch数。

File: paddlers/rs_models/cd/fccdn.py (hunk 4 above)
Comment (liuxtakeoff) [reply paddlers/rs_models/cd/fccdn.py]: 他命名逻辑是有两种Densecat，一个是cat_diff,就叫Densecat_cat_diff;一个是cat_add就叫Densecat_cat_add。似乎叫DenseCatDiff更简明些，晚上修改一下。

File: paddlers/rs_models/cd/__init__.py (hunk 6 above)
Comment (liuxtakeoff) [reply paddlers/rs_models/cd/__init__.py]: 但这样会不会无法体现这是fccdn里的损失，如果后续添加了其他自监督损失就无法区分了呀

File: paddlers/rs_models/cd/losses/fccdn_loss.py (hunk 16 above)
Comment (liuxtakeoff) [reply paddlers/rs_models/cd/losses/fccdn_loss.py:129]: 除了池化方法我没查找到其他resize tensor的方法。。resize方法好像是不能改变数据数目的，只能改变数据的形状，从[3,10,10] -> [3,20,5]这样，似乎做不到[3,10,10]->[3,5,5]。

File: paddlers/rs_models/cd/losses/fccdn_loss.py (hunk 8 above)
Comment (liuxtakeoff) [reply paddlers/rs_models/cd/losses/fccdn_loss.py]: 这个应该是作者写错了，我修改一下

File: paddlers/rs_models/cd/losses/fccdn_loss.py (hunk 9 above)
Comment (liuxtakeoff) [reply paddlers/rs_models/cd/losses/fccdn_loss.py]: 那是改为forward方法吗

File: paddlers/rs_models/cd/losses/fccdn_loss.py (hunk 16 above)
Comment (liuxtakeoff) [reply paddlers/rs_models/cd/losses/fccdn_loss.py:129]: 或许我可以加一个判断报错？如果labels和scores[0]尺寸不一样了，就报错提示用户“尺寸错误，请检查网络输入是否为16的倍数”

File: paddlers/rs_models/cd/__init__.py (hunk 6 above)
Comment (Bobholamovic) [reply paddlers/rs_models/cd/__init__.py]: 好的，这样的话请修改为`fccdn_ssl_loss`吧，命名上一般`loss`放最后。

File: paddlers/rs_models/cd/losses/fccdn_loss.py (hunk 9 above)
Comment (Bobholamovic) [reply paddlers/rs_models/cd/losses/fccdn_loss.py]: 是的

File: paddlers/rs_models/cd/losses/fccdn_loss.py (hunk 16 above)
Comment (Bobholamovic) [reply paddlers/rs_models/cd/losses/fccdn_loss.py:129]: 可以看一下`paddle.nn.functional.interpolate()`函数。

**Comment by liuxtakeoff** [comment]
对训练参数进行了调整，最终训练轮次为15，训练时间在10分钟之内。
train1
train2
鉴于可以节省的计算量不大，就没有创建mask记录索引值，其他都按要求进行了调整。

File: paddlers/rs_models/cd/fccdn.py [hunk 18]
```diff
@@ -361,10 +361,10 @@ class FCCDN(nn.Layer):
         (https://arxiv.org/pdf/2105.10860.pdf).
 
     Args:
-        in_channels(int): Number of input channels(default: 3).
+        in_channels (int): Number of input channels(default: 3).
```
Comment (Bobholamovic) [inline paddlers/rs_models/cd/fccdn.py]: `(default: ...)`这个括号也要注意空格。

File: paddlers/rs_models/cd/losses/fccdn_loss.py [hunk 19]
```diff
@@ -116,44 +110,48 @@ def __call__(self, scores, labels):
...
-    pred_seg_pre_tmp1 = paddle.ones(out1.shape)
-    pred_seg_pre_tmp1[out1 <= 0.5] = 0
-    pred_seg_post_tmp1 = paddle.ones(out2.shape)
-    pred_seg_post_tmp1[out2 <= 0.5] = 0
-
-    pred_seg_pre_tmp2 = paddle.ones(scores[0].shape)
-    pred_seg_pre_tmp2[out3 <= 0.5] = 0
-    pred_seg_post_tmp2 = paddle.ones(scores[1].shape)
-    pred_seg_post_tmp2[out4 <= 0.5] = 0
+    out1 = paddle.where(labels_downsample == 1, paddle.zeros_like(out1), out1)
```
Comment (Bobholamovic) [inline paddlers/rs_models/cd/losses/fccdn_loss.py:142]: 这里可以加一个断言，判断labels_downsample的dtype为integral类型，防止潜在错误。

File: test_tipc/configs/cd/fccdn/fccdn.yaml [hunk 20]
```diff
@@ -4,7 +4,12 @@ _base_: ../_base_/airchange.yaml
 
 save_dir: ./test_tipc/output/cd/fccdn/
 
-learning_rate: 0.03
+learning_rate: 0.07
+lr_decay_power: 0.6
+
```
Comment (Bobholamovic) [inline test_tipc/configs/cd/fccdn/fccdn.yaml]: 去除多余的空行，参考原来的yaml中的排布。

**Review by Bobholamovic — COMMENTED** [review]
精度验收通过，TIPC满足要求，代码基本满足规范，论文复现成功。再做一些格式上的小修改后可以合入PR。

File: paddlers/rs_models/cd/losses/fccdn_loss.py [hunk 21]
```diff
@@ -124,6 +124,9 @@ def fccdn_ssl_loss(scores, labels):
     # Get downsampled change map
     h, w = scores[0].shape[-2], scores[0].shape[-1]
     labels_downsample = F.interpolate(x=labels.unsqueeze(1), size=[h, w])
+    labels_type = str(labels_downsample.dtype)
+    assert "int" in labels_type or "bool" in labels_type,\
+        f"Expect labels type int or bool, but the type is {labels_type}"
```
Comment (Bobholamovic) [inline paddlers/rs_models/cd/losses/fccdn_loss.py]: f"Expected dtype of `labels` to be int or bool, but got {labels_type}."

File: paddlers/rs_models/cd/losses/fccdn_loss.py [hunk 22]
```diff
@@ -114,7 +114,7 @@ def fccdn_ssl_loss(scores, labels):
     """
     FCCDN change detection self-supervised learning loss
```
Comment (Bobholamovic) [inline paddlers/rs_models/cd/losses/fccdn_loss.py]: 输入看起来是logits而不是scores（因为后面又做了sigmoid）。另外因为这个API要暴露给用户，所以需要按照docstring的格式要求撰写注释。例如：
```
Self-supervised learning loss for change detection.

The original article refers to
    Pan Chen, et al., "FCCDN: Feature Constraint Network for VHR Image Change Detection"

Args:
    logits_list (list[paddle.Tensor]): Single-channel segmentation logit maps for each of the two temporal phases.
    labels (paddle.Tensor): Binary change labels.
```

**Review by Bobholamovic — COMMENTED** [review]
之前有些忽略的细节，请修改一下~

File: paddlers/rs_models/cd/losses/fccdn_loss.py (hunk 22 above)
Comment (liuxtakeoff) [reply paddlers/rs_models/cd/losses/fccdn_loss.py]: 已按要求进行修改~

**Review by Bobholamovic — APPROVED** [review]
LGTM.
